### PR TITLE
#1029 WIP: 1029_Integrating_LibXML2

### PR DESCRIPTION
**Pull request by e2l-su**
First try of integrating LibXML2 before changing IIDM version. It still lacks a few things
1 - Documentation and licenses for LibXML2 must be added
2 - Changes may be necessary in envDynawo.sh (specially for deploying)
3 - No patch is applied to the installation: are some needed ?
... and probably more ...

The changes are quoted with '#-wip' mark in CMakeLists.txt. You may commit changes, as they do not affect existing libraries. It only adds (unnecessary) LibXML2 to the project, until next step.

### CODE REVIEW

**Comment by rosiereflo** [comment]
Since this changes the behavior of the compilation in CXX11 we will not merge this MR until the whole development is ready.

File: dynawo/3rdParty/CMakeLists.txt [hunk 1]
```diff
@@ -904,12 +901,21 @@ else()
   set(LIBZIP_HOME ${install_dir})
 endif()
 
-# if ( CXX11_ENABLED )                                                                              # New part for IIDM-2
+#-wip- if ( CXX11_ENABLED )                                                                              # New part for IIDM-2
+
+# Minimum required for powsybl that sets that rule
+cmake_minimum_required(VERSION 3.12)
```
Comment (gautierbureau) [inline dynawo/3rdParty/CMakeLists.txt:907]: This call can't be placed here, it has to be at the very top of a top CMakeLists.txt

https://cmake.org/cmake/help/latest/command/cmake_minimum_required.html

File: dynawo/cmake/SetPatchCommands.cmake [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    set( PatchFile "${TmpDirName}/${projectName}/patch/common/${projectName}.patch")
+    if( EXISTS ${PatchFile} )
+        if (UNIX)
+            set( paquet_all_comp_patch
+                patch -p1 --forward --no-backup-if-mismatch -r /dev/null -i ${PatchFile} || echo -n
+                PARENT_SCOPE
+            )
+        else()
+            set( paquet_all_comp_patch
```
Comment (gautierbureau) [inline dynawo/cmake/SetPatchCommands.cmake:26]: paquet is french

File: dynawo/cmake/SetPatchCommands.cmake [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+# See AUTHORS.txt
+# All rights reserved.
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, you can obtain one at http://mozilla.org/MPL/2.0/.
+# SPDX-License-Identifier: MPL-2.0
+#
+# This file is part of Dynawo, an hybrid C++/Modelica open source time domain simulation tool for power systems.
+
+function( SetPatchCommands projectName)
```
Comment (gautierbureau) [inline dynawo/cmake/SetPatchCommands.cmake:11]: Please follow the coding style in the file (top CMakeLists.txt, even with the same keyword you are not consistent through out the file

Comment (gautierbureau) [reply dynawo/cmake/SetPatchCommands.cmake:11]: And if you define it here you could already replace the same code in the top CMakeLists with the call of this function

File: dynawo/3rdParty/CMakeLists.txt [hunk 4]
```diff
@@ -904,12 +901,21 @@ else()
...
 
-        # HERE IS THE NEW PART
+#-wip- else()                                                                                            # old part : legacy IIDM
+
+# Minimum required (for ExternalProject_Add)
+cmake_minimum_required(VERSION 3.9.6)
 
-# else()                                                                                            # old part : legacy IIDM
 set(XERCESC_HOME      CACHE PATH "Path where a Xerces-C installation already exists")               # Start install XERCESC
+# Xercesc est aussi mentionné REQUIRED depuis le script CMakeLists.txt principal. A tester avant de le retirer
```
Comment (gautierbureau) [inline dynawo/3rdParty/CMakeLists.txt:918]: no french

File: dynawo/cmake/SetPatchCommands.cmake [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    set( TmpDirName "$ENV{DYNAWO_HOME}/dynawo/3rdParty")
```
Comment (gautierbureau) [inline dynawo/cmake/SetPatchCommands.cmake:13]: tab is two spaces wide (and not real tab either)

File: dynawo/cmake/SetPatchCommands.cmake [hunk 6]
```diff
@@ -0,0 +1,60 @@
+# Copyright (c) 2015-2019, RTE (http://www.rte-france.com)
```
Comment (gautierbureau) [inline dynawo/cmake/SetPatchCommands.cmake:1]: For new files we should put 2020 and not 2019

File: dynawo/cmake/Support-LibXml2.cmake [hunk 7]
```diff
@@ -0,0 +1,85 @@
+# Copyright (c) 2015-2019, RTE (http://www.rte-france.com)
```
Comment (gautierbureau) [inline dynawo/cmake/Support-LibXml2.cmake:1]: LibXml2.cmake would be better for the name

File: dynawo/cmake/Support-LibXml2.cmake [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+#
+# This file is part of Dynawo, an hybrid C++/Modelica open source time domain simulation tool for power systems.
+
+cmake_minimum_required( VERSION 3.12)
+
+if ( NOT CMAKE_SCRIPT_MODE_FILE )
+    project( "Libxml2 support for Dynawo")
+endif ()
+
+include( $ENV{DYNAWO_HOME}/dynawo/cmake/cpu_count.cmake)
```
Comment (gautierbureau) [inline dynawo/cmake/Support-LibXml2.cmake:17]: renme in CPUCount.cmake

File: dynawo/cmake/Support-LibXml2.cmake (hunk 7 above)
Comment (gautierbureau) [reply dynawo/cmake/Support-LibXml2.cmake:1]: This file should not be located here but in 3rdParty/libxml2

File: dynawo/cmake/Support-LibXml2.cmake [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                            "CXX=${CMAKE_CXX_COMPILER}"
+                            "CFLAGS=${CMAKE_C_FLAGS}"
+                            "CXXFLAGS=${CMAKE_CXX_FLAGS}"
+
+        BUILD_COMMAND       make -j ${CPU_COUNT} all
+
+        INSTALL_COMMAND     make install
+    )
+
+    set_target_properties(
```
Comment (gautierbureau) [inline dynawo/cmake/Support-LibXml2.cmake:76]: why is this needed?

File: dynawo/cmake/cpu_count.cmake [hunk 10]
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2015-2019, RTE (http://www.rte-france.com)
```
Comment (gautierbureau) [inline dynawo/cmake/cpu_count.cmake:1]: Corresponding code should be deleted from top CMakeLists.txt and be replace by this call

File: dynawo/3rdParty/CMakeLists.txt [hunk 11]
```diff
@@ -904,12 +901,21 @@ else()
   set(LIBZIP_HOME ${install_dir})
 endif()
 
-# if ( CXX11_ENABLED )                                                                              # New part for IIDM-2
+#-wip- if ( CXX11_ENABLED )                                                                              # New part for IIDM-2
+
+# Minimum required for powsybl that sets that rule
+cmake_minimum_required(VERSION 3.12)
+
+include( "Support-LibXml2")
```
Comment (gautierbureau) [inline dynawo/3rdParty/CMakeLists.txt:909]: when the correspondign file is moved it should be something like this include(libxml2/LibXml2.cmake)

File: dynawo/cmake/Support-LibXml2.cmake [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    message( STATUS "found LibXML2 lib and includes : ${LIBXML2_VERSION_STRING} in ${LIBXML2_HOME}")
+
+else ()
+    include( ExternalProject)
+    ExternalProject_Add(
+                            ${paquet_name}
+        INSTALL_DIR         ${paquet_install_dir}
+
+        GIT_REPOSITORY      https://gitlab.gnome.org/GNOME/libxml2.git
+        GIT_TAG             master
```
Comment (gautierbureau) [inline dynawo/cmake/Support-LibXml2.cmake:52]: should be ${RequiredVersion}

**Review by gautierbureau — CHANGES_REQUESTED** [review]

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/3rdParty/CMakeLists.txt (hunk 1 above)
Comment (e2l-su) [reply dynawo/3rdParty/CMakeLists.txt:907]: I have read that. There is already a cmake_minimum_requirement at the very top. Nothing warns about the fact that requirement may change while processing, in fact requirement changes while treating other CMakeLists.txt by example. Writing it here makes it possible to change only one part of the big CMakeLists.txt. I tried it and did not notice any trouble.

File: dynawo/cmake/SetPatchCommands.cmake (hunk 2 above)
Comment (e2l-su) [reply dynawo/cmake/SetPatchCommands.cmake:26]: Yes, I did it at purpose because I do not want that my variables interfere with existing ones. It's only a name for variables. I could have written 'pa_name' instead of 'paquet_name'. Choosing it looks french makes it a lot easier for me to recognise my variable and not to mix with other ones. That can be changed, no problem.

File: dynawo/cmake/SetPatchCommands.cmake (hunk 3 above)
Comment (e2l-su) [reply dynawo/cmake/SetPatchCommands.cmake:11]: I do not see what you mean exactly. I want to follow the coding style. About that function, it is a new one in replacement of the one inside the very big CMakeLists.txt. It took me very much time to understand it. What I wish to do is to avoid making that file bigger.

File: dynawo/3rdParty/CMakeLists.txt (hunk 4 above)
Comment (e2l-su) [reply dynawo/3rdParty/CMakeLists.txt:918]: That is a mistake, sorry...

File: dynawo/cmake/SetPatchCommands.cmake (hunk 5 above)
Comment (e2l-su) [reply dynawo/cmake/SetPatchCommands.cmake:13]: OK, 'expand --tab=2' will do the job. I used --tab=4 until now.

File: dynawo/cmake/Support-LibXml2.cmake (hunk 8 above)
Comment (e2l-su) [reply dynawo/cmake/Support-LibXml2.cmake:17]: OK

File: dynawo/3rdParty/CMakeLists.txt [hunk 13]
```diff
@@ -8,9 +8,6 @@
 #
 # This file is part of Dynawo, an hybrid C++/Modelica open source time domain simulation tool for power systems.
 
-# Minimum required (for ExternalProject_Add)
-cmake_minimum_required(VERSION 3.9.6 FATAL_ERROR)
```
Comment (gautierbureau) [inline dynawo/3rdParty/CMakeLists.txt:12]: You say there is a top cmake_minimum_requirement but I see it is deleted here

File: dynawo/cmake/Support-LibXml2.cmake (hunk 9 above)
Comment (e2l-su) [reply dynawo/cmake/Support-LibXml2.cmake:76]: set_target_property ? I write that part instead of "include_directories( ${CMAKE_INSTALL_PREFIX}/include/libxml2)". It may be wrong since I'm beginning with cmake.

File: dynawo/3rdParty/CMakeLists.txt [hunk 14]
```diff
@@ -904,12 +901,21 @@ else()
   set(LIBZIP_HOME ${install_dir})
 endif()
 
-# if ( CXX11_ENABLED )                                                                              # New part for IIDM-2
+#-wip- if ( CXX11_ENABLED )                                                                              # New part for IIDM-2
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/3rdParty/CMakeLists.txt:904]: (and the comment)

File: dynawo/cmake/SetPatchCommands.cmake (hunk 3 above)
Comment (gautierbureau) [reply dynawo/cmake/SetPatchCommands.cmake:11]: I know it is a replacement but I don't see the code deleted in the top CMakeLists.txt

For the coding style, we don't use space after (, so function(foo),
also if(Test) and not if (Test) or if ( Test) as you use both

File: dynawo/cmake/SetPatchCommands.cmake (hunk 2 above)
Comment (gautierbureau) [reply dynawo/cmake/SetPatchCommands.cmake:26]: I know it's only a name but we code in english with english names for variables

File: dynawo/cmake/cpu_count.cmake (hunk 10 above)
Comment (e2l-su) [reply dynawo/cmake/cpu_count.cmake:1]: I'm not sure I understand. On github I only see one line... You mean maybe I can get rid of old codes ? I will do that when I will be sure the code is not used somewhere else.

File: dynawo/3rdParty/CMakeLists.txt (hunk 11 above)
Comment (e2l-su) [reply dynawo/3rdParty/CMakeLists.txt:909]: that may be discussed. Not all the 3rd parties have a directory in the CMAKE_SOURCE_DIR or whatever. All our "finders" like FindLibZIP.cmake are under the same directory 3rdParty/cmake. It's simplier to follow that same rule. Support-qqch.cmake is a "finder" that configures installation when it does not find, which is quite the same. Do you follow me ?

File: dynawo/cmake/Support-LibXml2.cmake (hunk 12 above)
Comment (e2l-su) [reply dynawo/cmake/Support-LibXml2.cmake:52]: I require at least one version (2.9), but if it is needed, I download latest one. If you want I can force a required version, that may be more safe. In that case, which one ? powsybl does not set one. The libxml2 library is already in Linux. A safer solution could be to "dnf install libxml2-devel" or "apt install libxml2-dev". Doing so we may have differences across systems...

File: dynawo/3rdParty/CMakeLists.txt (hunk 1 above)
Comment (gautierbureau) [reply dynawo/3rdParty/CMakeLists.txt:907]: cf my comment on the top cmake_minimum_required deletion

File: dynawo/3rdParty/CMakeLists.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (e2l-su) [reply dynawo/3rdParty/CMakeLists.txt:904]: Sorry, Florent told me where to put it: at the beginning of the MR 'WIP: 1029 blabla...'. If it works good it will block merges :)

File: dynawo/cmake/Support-LibXml2.cmake (hunk 7 above)
[the remark this answers is not in the document]
Comment (e2l-su) [reply dynawo/cmake/Support-LibXml2.cmake:1]: OK

File: dynawo/3rdParty/CMakeLists.txt (hunk 13 above)
Comment (e2l-su) [reply dynawo/3rdParty/CMakeLists.txt:12]: Absolutely right, it is my mistake !

Comment (e2l-su) [reply dynawo/3rdParty/CMakeLists.txt:12]: I have a new problem too, that I don't know why travis failed. DO you have an idea ?

Comment (rosiereflo) [reply dynawo/3rdParty/CMakeLists.txt:12]: The error is the following:
[ 29%] Performing configure step for 'libxml2'

You must have libtool installed to compile libxml.

Download the appropriate package for your distribution,

or see http://www.gnu.org/software/libtool

make[2]: *** [CMakeFiles/libxml2.dir/build.make:108: stamp_dir/libxml2-configure] Error 1

File: dynawo/cmake/cpu_count.cmake (hunk 10 above)
Comment (gautierbureau) [reply dynawo/cmake/cpu_count.cmake:1]: Yes I mean you copy/paste code in a new file but don't delete the old one present in the top CMakeLists

File: dynawo/cmake/Support-LibXml2.cmake (hunk 9 above)
Comment (gautierbureau) [reply dynawo/cmake/Support-LibXml2.cmake:76]: I think we will have to clarify what you want to do. I think the set is useless altogether

File: dynawo/3rdParty/CMakeLists.txt (hunk 11 above)
Comment (gautierbureau) [reply dynawo/3rdParty/CMakeLists.txt:909]: For me your support is not the same as a FindXXX in the cmake folder so it does not belong in the same folder

File: dynawo/cmake/Support-LibXml2.cmake (hunk 12 above)
Comment (gautierbureau) [reply dynawo/cmake/Support-LibXml2.cmake:52]: We need to download a release and not a master that is not stable. Powsybl requires 2.9 as a minimal version so that is the version we should install. The solution dnf or apt is only there for linux and will be the prefered way to install it on those os but for macos and windows we will use the code you juste developed

File: dynawo/3rdParty/CMakeLists.txt (hunk 13 above)
Comment (gautierbureau) [reply dynawo/3rdParty/CMakeLists.txt:12]: thanks for the digging in the log. it means we need to add a new requirement in the linux libraries. So I also need to update the docker for travis

Comment (gautierbureau) [reply dynawo/3rdParty/CMakeLists.txt:12]: I updated the Dockerfile so tomorrow we will be able to launch the tests again

File: dynawo/cmake/SetPatchCommands.cmake (hunk 3 above)
Comment (e2l-su) [reply dynawo/cmake/SetPatchCommands.cmake:11]: I will change, no problem.

File: dynawo/cmake/cpu_count.cmake (hunk 10 above)
Comment (e2l-su) [reply dynawo/cmake/cpu_count.cmake:1]: That is a very small part. I may still leave the old code only to be sure not to break any further include. That is why I change names of variables and functions (and why I'm interested in naming in french: to avoid conflicts while WIP). In actual CMakeLists files all variables are global, and since maybe used in files included from files included by... That explains it ?

File: dynawo/cmake/SetPatchCommands.cmake (hunk 3 above)
Comment (e2l-su) [reply dynawo/cmake/SetPatchCommands.cmake:11]: powsybl does not have our coding style

File: dynawo/cmake/cpu_count.cmake (hunk 10 above)
Comment (rosiereflo) [reply dynawo/cmake/cpu_count.cmake:1]: You can change the files as if it was the final version: this MR will not be merged until the full development is ready

File: dynawo/3rdParty/CMakeLists.txt (hunk 11 above)
Comment (e2l-su) [reply dynawo/3rdParty/CMakeLists.txt:909]: We may discuss upon that point. Using one single and existing directory makes it a lot easier, as it is the case for FindXXX. Easier since we don't create a new directory, easier for adding new SupportXXX since one may find it exactly in our "modules" directory. Existing libxxx directories under 3rdPartiy are only for patches. SupportXXX is very different from patches. In fact it is only a wrapper of FindXXX that generates target for installing when it does not find. I still thing it is similar, that's why I borrow the concept from the way cmake keeps its modules: choosing a unique name and putting all modules in one single directory.

File: dynawo/cmake/Support-LibXml2.cmake (hunk 12 above)
Comment (e2l-su) [reply dynawo/cmake/Support-LibXml2.cmake:52]: You are right. But powsybl does not set any requirement for LibXML2 version. I did it on my own, without any reason. Choosing a version we must say something like 2.9.4, but it will be our choice, arbitrary. My own opinion is to say LibXML2 is inside Gnome, so we must install the package libxml2-dev or devel from apt or dnf. It is as other libraries we use, like libblas, libaplack, libarchive or libexpat (examples). In that way we won't have trouble to garranty we use the same library powsybl does. That is an important choice to make since the actual code uses 'git clone': easy to deploy, but instable, I agree. I have 3 samples of Support-LibXML2.cmake that uses different way to install that part... My opinion is that following powsybl way to solve the question will be our best choice

Comment (e2l-su) [reply dynawo/cmake/Support-LibXml2.cmake:52]: On windows system, and MacIntosh, we will do exactly as it is for other packages like libblas... and the new libIIDM. There must be an indication on installing LibXML2 in powsybl on Windows (?)

File: dynawo/3rdParty/CMakeLists.txt (hunk 13 above)
Comment (e2l-su) [reply dynawo/3rdParty/CMakeLists.txt:12]: Thanks too. libtool is already part of linux requirements in dynawo (see Building Dynaωo from sources on Linux from dynawo.org)

Comment (rosiereflo) [reply dynawo/3rdParty/CMakeLists.txt:12]: I reran the build and still the same error.

File: dynawo/cmake/SetPatchCommands.cmake (hunk 3 above)
Comment (gautierbureau) [reply dynawo/cmake/SetPatchCommands.cmake:11]: I know. the two projects are independant

File: dynawo/3rdParty/CMakeLists.txt (hunk 13 above)
Comment (gautierbureau) [reply dynawo/3rdParty/CMakeLists.txt:12]: I will have a look at it
